File: runtime/watcher.py
from __future__ import annotations

import logging
import time
from collections.abc import Callable

from capture.base import ScreenCapture
from config import AppConfig
from input.base import KeyboardAdapter
from matching.template import find_template, load_template
from runtime.actions import execute_actions

logger = logging.getLogger(__name__)


class Watcher:
    def __init__(self, config: AppConfig, capture: ScreenCapture,
                 keyboard: KeyboardAdapter, should_stop: Callable[[], bool] = lambda: False,
                 can_execute: Callable[[], bool] = lambda: True,
                 pattern_guard: object | None = None) -> None:
        self._config = config
        self._capture = capture
        self._keyboard = keyboard
        self._should_stop = should_stop
        self._can_execute = can_execute
        self._pattern_guard = pattern_guard
        self._templates = {target.target_id: load_template(str(target.image)) for target in config.targets}
        self._visible = {target.target_id: False for target in config.targets}
        self._last_trigger = {target.target_id: 0.0 for target in config.targets}

    def run(self) -> None:
        logger.info("watcher started with %d target(s)", len(self._config.targets))
        while not self._should_stop():
            self.scan_once()
            time.sleep(self._config.scan_interval_ms / 1000)
        logger.info("watcher stopped")

    def scan_once(self) -> None:
        now = time.monotonic()
        if self._pattern_guard is not None and not self._pattern_guard.is_allowed(self._capture.capture_center()):
            for target in self._config.targets:
                self._visible[target.target_id] = False
            logger.info("center pattern is not allowed; actions blocked")
            return
        for target in self._config.targets:
            if not target.enabled:
                continue
            if not self._can_execute():
                self._visible[target.target_id] = False
                continue
            match = find_template(
                self._capture.capture(target.region),
                self._templates[target.target_id],
                target.threshold,
            )
            is_visible = match is not None
            was_visible = self._visible[target.target_id]
            self._visible[target.target_id] = is_visible
            if not is_visible or was_visible or now - self._last_trigger[target.target_id] < target.cooldown_ms / 1000:
                continue
            logger.info("target detected: id=%s score=%.4f x=%d y=%d", target.target_id, match.score, match.x, match.y)
            if execute_actions(target.actions, self._keyboard, self._should_stop):
                self._last_trigger[target.target_id] = now
                logger.info("actions executed: id=%s", target.target_id)
            else:
                logger.warning("actions interrupted: id=%s", target.target_id)

File: src/watcher.ts
import * as path from 'node:path';
import { captureCenter, captureRegion } from './capture.js';
import { findTemplate, loadTemplate } from './matcher.js';
import { executeActions } from './input.js';
import type { LoadedConfig } from './config.js';

const allowedCenterPatterns = ['니플레티', '카시우스'];
const centerThreshold = 0.92;

type TemplateImage = Awaited<ReturnType<typeof loadTemplate>>;

export class Watcher {
  config: LoadedConfig;
  rootDir: string;
  onLog: (message: string) => void;
  visible = new Map<string, boolean>();
  lastTrigger = new Map<string, number>();
  templates = new Map<string, TemplateImage>();
  centerPatterns: TemplateImage[] = [];
  stopped = false;
  paused = false;

  constructor(config: LoadedConfig, rootDir: string, onLog: (message: string) => void) {
    this.config = config;
    this.rootDir = rootDir;
    this.onLog = onLog;
    this.visible = new Map(config.targets.map((target) => [target.id, false]));
    this.lastTrigger = new Map(config.targets.map((target) => [target.id, 0]));
    this.templates = new Map();
    this.centerPatterns = [];
    this.stopped = false;
    this.paused = false;
  }

  async initialize(): Promise<void> {
    for (const target of this.config.targets) this.templates.set(target.id, await loadTemplate(target.imagePath));
    for (const name of allowedCenterPatterns) this.centerPatterns.push(await loadTemplate(path.join(this.rootDir, 'policy', 'center_patterns', `${name}.png`)));
  }

  async isCenterAllowed(): Promise<boolean> {
    const frame = await captureCenter(this.config.monitor);
    for (const pattern of this.centerPatterns) if (await findTemplate(frame, pattern, centerThreshold)) return true;
    return false;
  }

  async scanOnce(): Promise<void> {
    if (!(await this.isCenterAllowed())) {
      for (const target of this.config.targets) this.visible.set(target.id, false);
      return;
    }
    for (const target of this.config.targets) {
      if (!target.enabled || this.stopped) continue;
      if (this.paused) { this.visible.set(target.id, false); continue; }
      const match = await findTemplate(await captureRegion(this.config.monitor, target.region), this.templates.get(target.id) as TemplateImage, target.threshold);
      const isVisible = Boolean(match);
      const wasVisible = this.visible.get(target.id);
      this.visible.set(target.id, isVisible);
      if (!isVisible || wasVisible || Date.now() - (this.lastTrigger.get(target.id) as number) < target.cooldownMs) continue;
      this.onLog(`target detected: id=${target.id} score=${(match as NonNullable<Awaited<ReturnType<typeof findTemplate>>>).score.toFixed(4)} x=${(match as { x: number }).x} y=${(match as { y: number }).y}`);
      if (await executeActions(target.actions, () => this.stopped)) {
        this.lastTrigger.set(target.id, Date.now());
        this.onLog(`actions executed: id=${target.id}`);
      }
    }
  }

  async run(): Promise<void> {
    this.onLog(`watcher started with ${this.config.targets.length} target(s)`);
    while (!this.stopped) {
      try { await this.scanOnce(); } catch (error) { this.onLog(`watcher error: ${(error as Error).message}`); throw error; }
      await new Promise((resolve) => setTimeout(resolve, this.config.scanIntervalMs));
    }
    this.onLog('watcher stopped');
  }

  stop(): void { this.stopped = true; }
  togglePause(): boolean { this.paused = !this.paused; return this.paused; }
}
